### PR TITLE
Move shard API

### PR DESCRIPTION
**Pull request by agourlay**
This PR adds a new API to trigger the shard transfers explicitly. (https://github.com/qdrant/qdrant/issues/861) 

It contains a new consensus test showing that the transfer works in practice.

### CODE REVIEW

File: openapi/openapi-collections.ytt.yaml [hunk 1]
```diff
@@ -200,4 +200,26 @@ paths:
...
+        - cluster
+      summary: Update collection cluster setup
+      operationId: update_collection_cluster
+      requestBody:
+        description: Collection cluster update operations
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/ClusterOperations"
+      parameters:
```
Comment (generall) [inline openapi/openapi-collections.ytt.yaml:217]: it also requires definition for `{collection_name}` in `path`

**Comment by e-ivkov** [comment]
Will the user then select if he wants automatic or manual shard movement mode?

File: openapi/openapi-collections.ytt.yaml (hunk 1 above)
Comment (agourlay) [reply openapi/openapi-collections.ytt.yaml:217]: good catch :+1: 
https://github.com/qdrant/qdrant/pull/869/commits/c05f7a74840febc7e0b610a92e56af1f08bd3e97

**Comment by agourlay** [comment]
@e-ivkov I do not think we have a clear vision regarding manual vs automatic shard movement mode.
I think your idea makes sense, it could be a user settings at term. We could also propose to pin shards to certain peers.

In the absence of automatic shard re-balancing, this endpoint enables us to test the transfer feature at various scales or to perform admin tasks manually.

**Comment by agourlay** [comment]
I have added a basic report of ongoing shard transfers in the collection cluster info API (https://github.com/qdrant/qdrant/issues/866)
https://github.com/qdrant/qdrant/pull/869/commits/9bd2c0129796c8afc442208f64eb422dc3fcc124

I just wanted to validate that it whole works properly

```json
{
    "peer_id": 13493294820967747131,
    "shard_count": 2,
    "local_shards": [
        {
            "shard_id": 0,
            "points_count": 4
        }
    ],
    "remote_shards": [
        {
            "shard_id": 1,
            "peer_id": 15343882160508435617
        }
    ],
    "shard_transfers": [
        {
            "shard_id": 0,
            "to": 15343882160508435617
        }
    ]
}
```

File: lib/collection/src/collection.rs [hunk 2]
```diff
@@ -423,7 +427,7 @@ impl Collection {
             existing_temporary_shard.before_drop().await;
 
             match existing_temporary_shard {
-                Local(local_shard) => {
+                Shard::Local(local_shard) => {
```
Comment (agourlay) [inline lib/collection/src/collection.rs:430]: this was the only spot using the import

**Review by generall — APPROVED** [review]

File: src/common/collections.rs [hunk 3]
```diff
@@ -51,3 +57,61 @@ pub async fn do_get_collection_cluster(
...
+    operation: ClusterOperations,
+    dispatcher: &Dispatcher,
+    wait_timeout: Option<Duration>,
+) -> Result<bool, StorageError> {
+    if dispatcher.consensus_state().is_none() {
+        return Err(StorageError::BadRequest {
+            description: "Distributed mode disabled".to_string(),
+        });
+    }
+    let consensus_state = dispatcher.consensus_state().unwrap();
```
Comment (e-ivkov) [inline src/common/collections.rs:73]: This probably looks cleaner and no unwraps there :smile: 
```rust
let consensus_state = dispatcher.consensus_state() {
    Some(state) => state,
    None => return Err(..),
}
```

Comment (agourlay) [reply src/common/collections.rs:73]: Thanks, it does look better.
I will remember this trick :+1:

**Pull request merged**